src/main/java/org/assertj/core/internal/DeepDifference.java

```diff
@@ -443,6 +464,44 @@ private static boolean compareArrays(Object array1, Object array2, List<String>
   }
 
   /**
+   * Deeply compares two collections referenced by dualKey. This method attempts
+   * to quickly determine inequality by length, then if lengths match, in case of
+   * collection type is Set and there are passed no custom comparators, there is used
+   * comparison on hashcodes basis, otherwise each element from one collection is checked
+   * for existence in another one using 'deep' comparison.
+   */
+  private static <K, V> boolean compareUnorderedCollection(Collection<K> col1, Collection<V> col2,
+                                                           List<String> path, Deque<DualKey> toCompare,
+                                                           Set<DualKey> visited,
+                                                           Map<String, Comparator<?>> comparatorByPropertyOrField,
+                                                           TypeComparators comparatorByType) {
+    if (col1.size() != col2.size()) {
+      return false;
+    }
+
+    boolean hasNoCustomComparators = (comparatorByPropertyOrField == null || comparatorByPropertyOrField.isEmpty())
+      && (comparatorByType == null || comparatorByType.isEmpty());
+    if (hasNoCustomComparators && col1 instanceof Set) {
+      // this comparison is used for performance optimization reasons
+      return compareUnorderedCollectionByHashCodes(col1, col2, path, toCompare, visited);
+    }
+
+    Collection<V> col2Copy = new LinkedList<>(col2);
+    for (Object o1 : col1) {
+      Iterator<V> iterator = col2Copy.iterator();
+      while (iterator.hasNext()) {
+        Object o2 = iterator.next();
+        if (determineDifferences(o1, o2, path, comparatorByPropertyOrField, comparatorByType).isEmpty()) {
```

Comment:
I was thinking about adding smth like `.item` to the path. For example, at `#testUnorderedCollectionWithCustomComparatorsByFieldName `

    SetWrapper a = new SetWrapper(newLinkedHashSet(new Wrapper("one"), new Wrapper("two")));
    fieldComparators.put("set.o", ALWAY_EQUALS);

`"set.o"` points to `Wrapper` field values "one" and "two", but there is no way, to specify path to `Wrapper` itself, as path `"set"`, points to both `set` field from `SetWrapper` and its elements of type `Wrapper`.

But as there is a workaround for this, and this change is to break currently used paths (`"set.o"`  should be replaced with `"set.item.o"`), I just mentioned it as a comment.